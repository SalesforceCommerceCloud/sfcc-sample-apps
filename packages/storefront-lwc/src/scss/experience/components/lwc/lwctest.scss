.lwctest {
    color: chocolate;
    background-color: azure;
}
