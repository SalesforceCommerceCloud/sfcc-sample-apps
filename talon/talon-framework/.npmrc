# We have two registries we could use:
# Internal (private)
registry=https://nexus.soma.salesforce.com/nexus/content/groups/npm-all/

# Also grab @types NPMs from Nexus, SFCI cannot reach any public registry anyway
@types:registry=https://nexus.soma.salesforce.com/nexus/content/groups/npm-all/

@sfcc-dev:registry=https://registry.npmjs.org/

@lwc-platform:registry=https://npm.lwcjs.org/

# registery=https://registry.npmjs.org/

//registry.npmjs.org/:_authToken=f1557f75-416b-4e40-891c-8de68a3a1c83
