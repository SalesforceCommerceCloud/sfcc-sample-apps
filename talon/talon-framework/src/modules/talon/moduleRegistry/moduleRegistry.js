import * as lwc from '@lwc/engine';
import { assert, autoBind } from 'talon/utils';
import { getBasePath, getMode, getLocale } from 'talon/configProvider';
import { getResourceUrl, getViewModuleFullyQualifiedName } from 'talon-common';
import scopedModuleResolver from 'talon/scopedModuleResolver';

const NAMESPACE_ALIASES = {
    'lightning': 'interop'
};

/**
 * Module registry class.
 *
 * A single instance of it will be used throughout the app
 * and selected methods will be exported.
 *
 * We still export the class itself for testing purpose so that we can
 * create as many instances as needed.
 */
export class ModuleRegistry {
    registry = {};
    resourceUids;

    /**
     * Assert that the dependency exists in the registry.
     * @param {String} dependency - the name of the dependency
     */
    assertHasModule(dependency) {
        assert(this.hasModule(dependency), `Cannot resolve dependency '${dependency}'`);
    }

    /**
     * Add multiple modules to the registry.
     *
     * This function is different from addModule() in the sense that it takes
     * already exported modules rather than an exporter function for the modules.
     *
     * This function can be used with labels, LWC component, ES modules.
     *
     * This function is meant to be called by code generated by the framework.
     *
     * @param {Object} modulesByName - A map of modules to add to the registry,
     *                  keyed by module name and in which values are exported modules.
     */
    addModules(modulesByName) {
        Object.entries(modulesByName).forEach(([name, module]) => {
            this.registry[name] = module;
        });
    }

    /**
     * Add a single module to the registry.
     *
     * This function is meant to be called by code generated by the framework.
     *
     * Modules should only exported/evaluated once, the call is ignored if the module
     * is already registered.
     *
     * @param {string} name - The name of the module to add
     * @param {string[]} dependencies - The list of dependencies to pass to the exporter function
     * @param {Function} exporter - A function that will export the module to add.
     */
    define(name, dependencies, exporter) {
        this.addModule(null, name, dependencies, exporter);
    }

    /**
     * Add a single module to the registry.
     *
     * This function is meant to be called by code generated by the framework.
     *
     * Modules should only exported/evaluated once, the call is ignored if the module
     * is already registered.
     *
     * @param {string} descriptor - Not used, kept here for Aura compatibility reasons
     * @param {string} name - The name of the module to add
     * @param {string[]} dependencies - The list of dependencies to pass to the exporter function
     * @param {Function} exporter - A function that will export the module to add.
     */
    addModule(descriptor, name, dependencies, exporter) {
        if (exporter === undefined && typeof dependencies === 'function') {
            // amd define does not include dependencies param if no dependencies.
            this.addModule(descriptor, name, [], dependencies);
            return;
        }

        // ignore if module is already registered
        if (this.registry[name]) {
            return;
        }

        const moduleExports = {};
        this.registry[name] = exporter.apply(undefined, dependencies.map(dependency => {
            if (name === dependency) {
                return this.evaluateCircularDependency(name);
            }

            return this.evaluateModuleDependency(dependency, moduleExports);
        })) || moduleExports;
    }

    /**
     * Returns a Proxy delegating to the module from the registry
     * with the specified name.
     *
     * This is useful for circular dependencies when the module is not in the
     * registry yet at the time to evaluate it.
     *
     * @param {*} name The name of the module
     */
    evaluateCircularDependency(name) {
        const registry = this.registry;
        return new Proxy({}, {
            get(obj, prop) {
                return registry[name][prop];
            }
        });
    }

    /**
     * Evaluate module dependency from its full import name.
     *
     * eg 'lwc' or 'lightning/button' or '@salesforce/cssvars/customProperties'
     *
     * @param {string} dependency - A dependency name
     * @param {string} moduleExports - The dependency's exports
     */
    evaluateModuleDependency(dependency, moduleExports) {
        // Found itself
        if (this.registry[dependency]) {
            return this.registry[dependency];
        }

        // Handle special cases
        if (dependency === 'lwc') {
            return lwc;
        } else if (dependency === 'exports') {
            return moduleExports;
        }

        if (dependency.startsWith('@')) {
            // Handle scoped modules
            this.registry[dependency] = scopedModuleResolver.resolve(dependency);
        } else {
            // Handle unscoped case
            const [moduleName, component] = dependency.split('/');
            this.registry[dependency] = this.evaluateUnscopedModuleDependency(moduleName, component);
        }

        this.assertHasModule(dependency);
        return this.registry[dependency];
    }

    /**
     * Evaluate unscoped dependency from its module name and component.
     *
     * eg 'lightning/button' or 'interop/menuItem'
     *
     * @param {string} moduleName - An unscoped module
     * @param {string} component - The component name
     */
    evaluateUnscopedModuleDependency(moduleName, component) {
        if (NAMESPACE_ALIASES[moduleName]) {
            const aliasedName = [NAMESPACE_ALIASES[moduleName], component].join('/');
            this.assertHasModule(aliasedName);
            return this.registry[aliasedName];
        }

        throw new Error(`Cannot resolve module '${moduleName}'`);
    }

    /**
     * Gets a generated view template from the registry, loading it from the server if needed
     *
     * @param {string} name - The template name
     * @returns a promise which resolves the exported module
     */
    async getTemplate(name) {
        return this.getModule(getViewModuleFullyQualifiedName(name), this.getResourceUrl({ view: name }));
    }

    /**
     * Gets a generated component from the registry, loading it from the server if needed
     *
     * @param {string} name - The component name
     * @returns a promise which resolves the exported module
     */
    async getComponent(name) {
        return this.getModule(name, this.getResourceUrl({ component: name }));
    }

    /**
     * Gets a module synchronously from the registry if it is present.
     *
     * @param {string} name - The module name
     * @returns the exported module if present in the registry, null if not
     */
    getModuleIfPresent(name) {
        return this.registry[name];
    }

    /**
     * Gets a module from the registry, loading it from the server if needed.
     *
     * @param {string} name - The module name
     * @returns a promise which resolves the exported module
     */
    async getModule(name, resourceUrl) {
        let moduleFromRegistry = this.registry[name];

        // return the module from the registry
        if (moduleFromRegistry) {
            return moduleFromRegistry;
        }

        // fetch the component from the server if it is not available yet
        return new Promise((resolve, reject) => {
            const script = document.createElement("script");
            script.type = "text/javascript";

            script.src = `${getBasePath()}${resourceUrl || this.getResourceUrl({ component: name })}`;
            script.onload = () => {
                script.onload = script.onerror = undefined;
                moduleFromRegistry = this.registry[name];
                if (moduleFromRegistry) {
                    resolve(moduleFromRegistry);
                } else {
                    reject(`Failed to load module: ${name}`);
                }
            };
            script.onerror = (error) => {
                script.onload = script.onerror = undefined;
                reject(error);
            };
            document.body.appendChild(script);
        });
    }

    setResourceUids(resourceUids) {
        this.resourceUids = resourceUids;
    }

    getResourceUrl({ component, view }) {
        const prefix = component ? "component" : "view";
        const resource = component || view;
        const mode = getMode();
        const { langLocale } = getLocale();
        const resourceName = `${prefix}://${resource}@${langLocale}`;
        const uid = this.resourceUids && this.resourceUids[resourceName];
        return getResourceUrl(resourceName, mode, uid);
    }

    hasModule(name) {
        const module = this.registry[name];
        return (typeof module !== 'undefined' && module !== null);
    }
}

// create an instance with bound methods so that they can be exported
const instance = autoBind(new ModuleRegistry());

export const { addModule, addModules, getModule, getComponent, getTemplate, getModuleIfPresent, hasModule, setResourceUids, define } = instance;

export default { addModule, addModules, getModule, getComponent, getTemplate, getModuleIfPresent, hasModule, setResourceUids, define };
