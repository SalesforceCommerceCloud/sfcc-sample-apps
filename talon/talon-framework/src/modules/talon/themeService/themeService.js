import { autoBind } from "talon/utils";

export class ThemeService {
    theme;
    viewToThemeLayoutMap;

    /**
     *
     * @param {Object} theme - The application's main theme
     */
    setTheme(theme) {
        this.theme = theme;
    }

    /**
     * Get a theme layout component by type
     *
     * @param {String} type - The theme layout type to get
     */
    getThemeLayoutByType(type) {
        if (!this.theme.themeLayouts.hasOwnProperty(type)) {
            throw new Error(`No theme layout type by the name "${type}" found.`);
        }
        const themeLayout = this.theme.themeLayouts[type];
        return themeLayout.component || themeLayout.view;
    }

    /**
     * Get a theme layout by its view
     */
    getThemeLayoutByView(view) {
        if (!this.viewToThemeLayoutMap.hasOwnProperty(view)) {
            throw new Error(`No theme layout matching the "${view}" view.`);
        }
        return this.viewToThemeLayoutMap[view];
    }

    /**
     * Sets the view to theme layout map
     * @param {*} map - The map
     */
    setViewToThemeLayoutMap(map) {
        this.viewToThemeLayoutMap = map;
    }
}

// create an instance with bound methods so that they can be exported
const instance = autoBind(new ThemeService());

export const { setTheme, getThemeLayoutByType, setViewToThemeLayoutMap, getThemeLayoutByView } = instance;

export default { setTheme, getThemeLayoutByType, setViewToThemeLayoutMap, getThemeLayoutByView };