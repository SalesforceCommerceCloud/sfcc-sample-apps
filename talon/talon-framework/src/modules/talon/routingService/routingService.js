import { getBasePath } from "talon/configProvider";
import { assert, autoBind, getDefinedValues, getQueryStringParameters, mapToQueryString } from "talon/utils";
import { getThemeLayoutByView } from "talon/themeService";
import pathToRegexp from 'path-to-regexp';

const WHITELISTED_PARAMETERS = ['mode', 'talon.lwc.fallback'];

/**
 * Routing service class.
 *
 * A single instance of it will be used throughout the app
 * and selected methods will be exported.
 *
 * We still export the class itself for testing purpose so that we can
 * create as many instances as needed.
 */
export class RoutingService {
    router;
    routesByName = {};
    observers = new Set();
    currentRoute;
    currentParams = {};
    currentQueryParams = {};

    /**
     * Register a set of routes and start the router.
     *
     * See JSON schema at https://salesforce.quip.com/EDbbAKBI92ZM#LeeACA0AB7l.
     *
     * See validate-routes.js#validateRoutes() for the validation rules.
     *
     * @param {Object[]} routes - The routes to register
     */
    registerRoutes(routes) {
        assert(this.router, "Router implementation not set");

        routes.forEach((route) => {
            // save route for lookup by name
            this.routesByName[route.name] = route;

            const callback = this.onRouteChange.bind(this, route.name);

            if (route.isDefault) {
                this.router('*', callback);
            } else {
                this.router(route.path, callback);
            }
        });

        // set base path
        const basePath = getBasePath();
        if (basePath) {
            this.router.base(basePath);
        }

        // start the router
        this.router.start();
    }

    /**
     * Get the URL for a given route.
     * This can be used to populate links href.
     *
     * The returned URL will include any whitelisted parameters
     * specified in the current URL.
     *
     * @param {string} name - The route name
     * @param {object} routeParams - The route params
     * @param {object} queryParams - The query params
     */
    getRouteUrl(name, routeParams = {}, queryParams = {}) {
        assert(this.routesByName[name], `Unknown route: ${name}`);

        let routeUrl = getBasePath() + (this.routesByName[name] || {}).path;
        routeUrl = this.injectRouteParams(routeUrl, routeParams);

        // add present whitelisted parameters to the route URL
        const stickyQueryParams = WHITELISTED_PARAMETERS.reduce((acc, param) => {
            if (this.currentQueryParams[param]) {
                acc[param] = this.currentQueryParams[param];
            }
            return acc;
        }, {});
        // the requested query params override any current whitelisted params
        const combinedQueryParams = Object.assign({}, stickyQueryParams, queryParams);

        if (Object.keys(combinedQueryParams).length > 0) {
            return routeUrl + "?" + mapToQueryString(combinedQueryParams, true);
        }

        return routeUrl;
    }

    injectRouteParams(path, params = {}) {
        // TODO handle path-to-regexp failure (issue #304)
        const toPath = pathToRegexp.compile(path);
        return toPath(params);
    }

    /**
     * Returns a route which has been registered
     *
     * @param {string} name
     */
    getRoute(name) {
        assert(this.routesByName[name], `Unknown route: ${name}`);
        return this.routesByName[name] || {};
    }

    /**
     * Callback invoked by the router when route has changed.
     *
     * Makes sure to get the relevant information for all observers
     *
     * @param {string} name the target route name
     * @param {string} routeContext the route context
     */
    onRouteChange(name, routeContext = {}) {
        const route = this.routesByName[name];

        // the app container and the router container so that it can create and render
        // the page component and the theme layout
        this.currentRoute = route;
        this.themeLayout = getThemeLayoutByView(route.view);

        this.currentQueryParams = getQueryStringParameters(routeContext.querystring || '');
        const currentRouteParams = getDefinedValues(routeContext.params || {});
        // Route params always take precedence over query string params
        this.currentParams = Object.assign({}, this.currentQueryParams, currentRouteParams);
        this.observers.forEach(observer => {
            observer.next(route, this.currentParams);
        });
    }

    /**
     * Navigates to the route with the given name.
     *
     * @param {string} name - The route name
     * @param {object} params - Any route params to pass in
     */
    navigateToRoute(name, params = {}, queryParams = {}) {
        this.router.show(this.getRouteUrl(name, params, queryParams));
    }

    /**
     * Subscribe to route changes.
     *
     * @param {Object} observer - An observer that will be notified on route change.
     *                      Observers are objects with a next() method that will be invoked
     *                      with the new route every time the route changes.
     * @returns A subscription object with an unsubscribe() method that can be used to stop
     *          receiving notifications.
     * @example
     *  const subscription = subscribe({
     *      next: (route) => {
     *          // handle route change
     *      }
     *  });
     *
     *  // later
     *  subscription.unsubscribe();
     *
     */
    subscribe(observer) {
        if (this.currentRoute) {
            observer.next(this.currentRoute, this.currentParams);
        }

        this.observers.add(observer);

        return {
            unsubscribe: () => {
                this.observers.delete(observer);
            }
        };
    }

    /**
     * Sets the 3rd party router, currently https://visionmedia.github.io/page.js/.
     * We do this so that we can compile in core where the 3rd party router is not available.
     *
     * Also resets current routes and observer since they are tied to the current router
     * although setting router impl multiple times should only happen in tests.
     *
     * @param {Object} routerImpl - 3rd party router
     */
    setRouter(routerImpl) {
        this.router = routerImpl;
    }
}

// create an instance with bound methods so that they can be exported
const instance = autoBind(new RoutingService());

export const { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter } = instance;

export default { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter };