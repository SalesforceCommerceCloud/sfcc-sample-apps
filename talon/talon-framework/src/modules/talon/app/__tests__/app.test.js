import App  from 'talon/app';
import { createElement } from 'lwc';
import { subscribe } from 'talon/routingService';
import { default as mockTemplate } from './fixtures/themeLayout';
import { getTemplate } from 'talon/moduleRegistry';
import { getThemeLayoutByView } from 'talon/themeService';

beforeEach(() => {
    while (document.body.firstChild) {
        document.body.removeChild(document.body.firstChild);
    }

    jest.resetAllMocks();

    subscribe.mockReturnValue({
        unsubscribe: jest.fn()
    });

    getThemeLayoutByView.mockReturnValue("themeLayout");
    getTemplate.mockReturnValue(Promise.resolve({html: mockTemplate.html, attributes: mockTemplate.attributes}));
});

jest.mock('talon/moduleRegistry', () => {
    return {
        getTemplate: jest.fn()
    };
});

jest.mock('talon/themeService', () => {
    return {
        getThemeLayoutByView: jest.fn()
    };
});

jest.mock('talon/routingService', () => {
    return {
        subscribe: jest.fn()
    };
});

describe('talon/app', () => {
    it('renders no theme layout', async () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        expect(element).toMatchSnapshot();
    });

    it('renders the theme layout', () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" });

        return new Promise(window.requestAnimationFrame)
            .then(() => {
                expect(element).toMatchSnapshot();
            });
    });

    it('renders the theme layout with empty route params', () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" }, {});

        return new Promise(window.requestAnimationFrame)
            .then(() => {
                expect(element).toMatchSnapshot();
            });
    });

    it('renders the theme layout with route params', () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" }, {recordId: "ABC123"});

        return new Promise(window.requestAnimationFrame)
            .then(() => {
                expect(element).toMatchSnapshot();
            });
    });

    it('does not fetch the template if the theme layout hasnt changed', async () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" }, {recordId: "ABC123"});

        return Promise.resolve().then(() => {
            expect(getTemplate).toHaveBeenCalledTimes(1);
            observer.next({ view: "inner" }, {recordId: "ABC123"});
            return new Promise(window.requestAnimationFrame)
                .then(() => {
                    expect(getTemplate).toHaveBeenCalledTimes(1);
                });
        });
    });

    it('route params are updated even if the theme layout hasnt changed', async () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" }, {recordId: "ABC123"});

        return Promise.resolve().then(() => {
            expect(getTemplate).toHaveBeenCalledTimes(1);
            observer.next({ view: "inner" }, {recordId: "FOOBAR"});
            return new Promise(window.requestAnimationFrame)
                .then(() => {
                    expect(getTemplate).toHaveBeenCalledTimes(1);
                    expect(element).toMatchSnapshot();
                });
        });
    });

    it('subscribes to route changes', async () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        expect(subscribe).toHaveBeenCalledTimes(1);
    });

    it('unsubscribes when disconnected', async () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);
        document.body.removeChild(element);

        const { unsubscribe } = subscribe.mock.results[0].value;
        expect(unsubscribe).toHaveBeenCalledTimes(1);
    });

    it('sets the right route params when theme layout changes', () => {
        const element = createElement('talon-app', { is: App });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "inner" }, {recordId: "ABC123"});

        return Promise.resolve().then(() => {
            expect(getTemplate).toHaveBeenCalledTimes(1);
            observer.next({ view: "default" }, {recordId: "FOOBAR"});
            return new Promise(window.requestAnimationFrame)
                .then(() => {
                    expect(getTemplate).toHaveBeenCalledTimes(1);
                    expect(element).toMatchSnapshot();
                });
        });
    });
});