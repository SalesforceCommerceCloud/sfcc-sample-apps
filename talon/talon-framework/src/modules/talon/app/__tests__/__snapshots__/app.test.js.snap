// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`talon/app renders no theme layout 1`] = `
<talon-app>
  #shadow-root(open)
</talon-app>
`;

exports[`talon/app renders the theme layout 1`] = `
<talon-app>
  #shadow-root(open)
    <h1>
      Theme Layout: 
    </h1>
    <p>
      What kind of stuff? 
      Hello
       
    </p>
</talon-app>
`;

exports[`talon/app renders the theme layout with empty route params 1`] = `
<talon-app>
  #shadow-root(open)
    <h1>
      Theme Layout: 
    </h1>
    <p>
      What kind of stuff? 
      Hello
       
    </p>
</talon-app>
`;

exports[`talon/app renders the theme layout with route params 1`] = `
<talon-app>
  #shadow-root(open)
    <h1>
      Theme Layout: 
      ABC123
    </h1>
    <p>
      What kind of stuff? 
      Hello
       
    </p>
</talon-app>
`;

exports[`talon/app route params are updated even if the theme layout hasnt changed 1`] = `
<talon-app>
  #shadow-root(open)
    <h1>
      Theme Layout: 
      FOOBAR
    </h1>
    <p>
      What kind of stuff? 
      Hello
       
    </p>
</talon-app>
`;

exports[`talon/app sets the right route params when theme layout changes 1`] = `
<talon-app>
  #shadow-root(open)
    <h1>
      Theme Layout: 
      FOOBAR
    </h1>
    <p>
      What kind of stuff? 
      Hello
       
    </p>
</talon-app>
`;
